File: apps/api/src/routes/ordem-items.ts
import { Hono } from 'hono'
import * as XLSX from 'xlsx'
import { prisma, type Prisma } from '@qtscout/db'
import { canManageItem, resolveCategory, type ProfileForAuth } from '@qtscout/core/ordem-permissions'
import { annotateItems, resolveRefs } from '@qtscout/core/ordem-resolver'
import { mapActivityRow, type ActivityPayload, type ActivityRow } from '@qtscout/core/siie-atividades-import'
import { isOrdemSection, validateItemData, type ItemShape, type OrdemSection } from '@qtscout/types/ordem-item'
import { requireAuth, requireAdmin } from '../middleware/auth'
import type { AppEnv } from '../types'

async function validateRefs(
  shape: ItemShape,
  value: Record<string, unknown>,
  section: OrdemSection | null,
): Promise<string | null> {
  const scoutIds: string[] = []
  const profileIds: string[] = []

  if (shape === 'MEMBER_REF' && typeof value.scoutId === 'string') {
    scoutIds.push(value.scoutId)
  }
  if (shape === 'NOITES_REF' && Array.isArray(value.scoutIds)) {
    for (const id of value.scoutIds) if (typeof id === 'string') scoutIds.push(id)
  }
  if (shape === 'PROFILE_REF' && typeof value.profileId === 'string') {
    profileIds.push(value.profileId)
  }
  if (shape === 'SCOUT_OR_PROFILE_REF' && typeof value.refId === 'string') {
    if (value.kind === 'scout') scoutIds.push(value.refId)
    if (value.kind === 'profile') profileIds.push(value.refId)
  }

  if (scoutIds.length > 0) {
    const found = await prisma.scout.findMany({
      where: { id: { in: scoutIds } },
      select: { id: true, section: true },
    })
    if (found.length !== scoutIds.length) return 'Membro não encontrado'
    if (section) {
      const mismatched = found.find((s) => s.section !== section)
      if (mismatched) return 'O membro selecionado pertence a outra secção'
    }
  }
  if (profileIds.length > 0) {
    const count = await prisma.profile.count({ where: { id: { in: profileIds } } })
    if (count !== profileIds.length) return 'Dirigente não encontrado'
  }

  return null
}

async function loadAuth(userId: string): Promise<ProfileForAuth | null> {
  const p = await prisma.profile.findUnique({
    where: { id: userId },
    select: { role: true, roles: true, section: true },
  })
  return p ? { role: p.role, roles: p.roles, section: p.section } : null
}

export const ordemItems = new Hono<AppEnv>()
ordemItems.use('*', requireAuth)

ordemItems.get('/', async (c) => {
  const from = c.req.query('from')
  const to = c.req.query('to')
  const section = c.req.query('section')
  const category = c.req.query('category')
  const included = c.req.query('included') // 'true' | 'false' | undefined

  const where: Prisma.OrdemItemWhereInput = {}
  if (from || to) {
    where.date = {}
    if (from) where.date.gte = new Date(from)
    if (to) where.date.lte = new Date(to)
  }
  if (section && isOrdemSection(section)) where.section = section
  if (category) where.category = category
  if (included === 'true') where.includedInOsId = { not: null }
  if (included === 'false') where.includedInOsId = null

  const items = await prisma.ordemItem.findMany({
    where,
    orderBy: [{ date: 'desc' }, { createdAt: 'desc' }],
    include: { createdBy: { select: { id: true, name: true, email: true } } },
  })

  const refs = await resolveRefs(items)
  const annotated = annotateItems(items, refs)
  return c.json({ items: annotated })
})

ordemItems.post('/', async (c) => {
  const session = c.get('session')
  const body = (await c.req.json()) as {
    category?: unknown; section?: unknown; date?: unknown; data?: unknown
  }

  const category = typeof body.category === 'string' ? resolveCategory(body.category) : null
  if (!category) return c.json({ error: 'Categoria inválida' }, 400)

  const section =
    body.section == null ? null : isOrdemSection(body.section) ? body.section : 'invalid'
  if (section === 'invalid') return c.json({ error: 'Secção inválida' }, 400)
  if (category.scope === 'SECTION' && !section) {
    return c.json({ error: 'Secção é obrigatória' }, 400)
  }
  if (category.scope === 'GROUP' && section) {
    return c.json({ error: 'Categoria de grupo não aceita secção' }, 400)
  }

  if (typeof body.date !== 'string' || isNaN(Date.parse(body.date))) {
    return c.json({ error: 'Data inválida' }, 400)
  }

  const dataResult = validateItemData(category.shape, body.data)
  if (!dataResult.ok) return c.json({ error: dataResult.error }, 400)

  const profile = await loadAuth(session.user.id)
  if (!profile) return c.json({ error: 'Perfil não encontrado' }, 404)

  if (!canManageItem(profile, category, section)) {
    return c.json({ error: 'Sem permissões para esta categoria/secção' }, 403)
  }

  const refError = await validateRefs(category.shape, dataResult.value, section)
  if (refError) return c.json({ error: refError }, 400)

  const item = await prisma.ordemItem.create({
    data: {
      category: category.key,
      section,
      date: new Date(body.date),
      data: dataResult.value as Prisma.InputJsonValue,
      createdById: session.user.id,
    },
    include: { createdBy: { select: { id: true, name: true, email: true } } },
  })

  return c.json({ item })
})

// SIIE activity import (multipart). Registered before /:id so the static path wins.
ordemItems.post('/import-activities', requireAdmin, async (c) => {
  const session = c.get('session')

  const formData = await c.req.formData()
  const file = formData.get('file')
  if (!(file instanceof File)) {
    return c.json({ error: 'Ficheiro não fornecido' }, 400)
  }

  let rows: ActivityRow[]
  try {
    const buf = Buffer.from(await file.arrayBuffer())
    const wb = XLSX.read(buf, { type: 'buffer', cellDates: true })
    const sheet = wb.Sheets[wb.SheetNames[0]]
    if (!sheet) return c.json({ error: 'Folha vazia' }, 400)
    rows = XLSX.utils.sheet_to_json<ActivityRow>(sheet, { raw: true, defval: null })
  } catch (err) {
    return c.json({ error: `Não foi possível ler o ficheiro: ${(err as Error).message}` }, 400)
  }

  type ImportSummary = {
    total: number; created: number; updated: number; skipped: number
    errors: { row: number; descricao: string; error: string }[]
  }
  type ValidRow = { row: number; payload: ActivityPayload }

  const summary: ImportSummary = { total: rows.length, created: 0, updated: 0, skipped: 0, errors: [] }

  const valid: ValidRow[] = []
  for (let i = 0; i < rows.length; i++) {
    const mapped = mapActivityRow(rows[i], i + 2)
    if (!mapped.ok) {
      summary.errors.push({ row: mapped.row, descricao: mapped.descricao, error: mapped.error })
      continue
    }
    valid.push({ row: i + 2, payload: mapped.payload })
  }

  if (valid.length === 0) return c.json({ summary })

  const existing = await prisma.ordemItem.findMany({
    where: { externalId: { in: valid.map((v) => v.payload.externalId) } },
    select: { externalId: true, includedInOsId: true },
  })
  const existingMap = new Map(existing.map((e) => [e.externalId as string, e.includedInOsId]))

  type UpsertOutcome =
    | { ok: true; row: number; wasExisting: boolean; skipped: boolean }
    | { ok: false; row: number; descricao: string; error: string }

  const outcomes = await Promise.all(
    valid.map(async ({ row, payload }): Promise<UpsertOutcome> => {
      const includedInOsId = existingMap.get(payload.externalId)
      const wasExisting = existingMap.has(payload.externalId)
      if (wasExisting && includedInOsId) {
        return { ok: true, row, wasExisting: true, skipped: true }
      }
      const data: Prisma.InputJsonValue = {
        nome: payload.nome,
        datas: payload.datas,
        local: payload.local,
      }
      try {
        await prisma.ordemItem.upsert({
          where: { externalId: payload.externalId },
          create: {
            externalId: payload.externalId,
            category: 'ATIVIDADE',
            section: payload.section,
            date: payload.date,
            data,
            createdById: session.user.id,
          },
          update: {
            category: 'ATIVIDADE',
            section: payload.section,
            date: payload.date,
            data,
          },
        })
        return { ok: true, row, wasExisting, skipped: false }
      } catch (err) {
        const e = err as { code?: string; message?: string }
        return {
          ok: false, row, descricao: payload.nome,
          error: e.code === 'P2002' ? 'Conflito de unicidade' : e.message ?? 'Erro desconhecido',
        }
      }
    }),
  )

  for (const o of outcomes) {
    if (!o.ok) {
      summary.errors.push({ row: o.row, descricao: o.descricao, error: o.error })
      continue
    }
    if (o.skipped) summary.skipped++
    else if (o.wasExisting) summary.updated++
    else summary.created++
  }

  return c.json({ summary })
})

ordemItems.patch('/:id', async (c) => {
  const session = c.get('session')
  const id = c.req.param('id')

  const existing = await prisma.ordemItem.findUnique({ where: { id } })
  if (!existing) return c.json({ error: 'Item não encontrado' }, 404)
  if (existing.includedInOsId) {
    return c.json({ error: 'Item já incluído numa Ordem de Serviço' }, 409)
  }

  const category = resolveCategory(existing.category)
  if (!category) return c.json({ error: 'Categoria inválida' }, 500)

  const profile = await loadAuth(session.user.id)
  if (!profile) return c.json({ error: 'Perfil não encontrado' }, 404)

  const isOwner = existing.createdById === session.user.id
  if (!isOwner && profile.role !== 'ADMIN') {
    return c.json({ error: 'Sem permissões' }, 403)
  }

  const body = (await c.req.json()) as { date?: unknown; data?: unknown; section?: unknown }

  const update: Prisma.OrdemItemUpdateInput = {}
  if (body.date !== undefined) {
    if (typeof body.date !== 'string' || isNaN(Date.parse(body.date))) {
      return c.json({ error: 'Data inválida' }, 400)
    }
    update.date = new Date(body.date)
  }
  if (body.data !== undefined) {
    const r = validateItemData(category.shape, body.data)
    if (!r.ok) return c.json({ error: r.error }, 400)
    update.data = r.value as Prisma.InputJsonValue
  }
  if (body.section !== undefined) {
    if (category.scope === 'GROUP') {
      return c.json({ error: 'Categoria de grupo não aceita secção' }, 400)
    }
    if (!isOrdemSection(body.section)) {
      return c.json({ error: 'Secção inválida' }, 400)
    }
    if (!canManageItem(profile, category, body.section)) {
      return c.json({ error: 'Sem permissões para essa secção' }, 403)
    }
    update.section = body.section
  }

  const item = await prisma.ordemItem.update({
    where: { id },
    data: update,
    include: { createdBy: { select: { id: true, name: true, email: true } } },
  })
  return c.json({ item })
})

ordemItems.delete('/:id', async (c) => {
  const session = c.get('session')
  const id = c.req.param('id')

  const existing = await prisma.ordemItem.findUnique({ where: { id } })
  if (!existing) return c.json({ error: 'Item não encontrado' }, 404)
  if (existing.includedInOsId) {
    return c.json({ error: 'Item já incluído numa Ordem de Serviço' }, 409)
  }

  const profile = await loadAuth(session.user.id)
  if (!profile) return c.json({ error: 'Perfil não encontrado' }, 404)
  const isOwner = existing.createdById === session.user.id
  if (!isOwner && profile.role !== 'ADMIN') {
    return c.json({ error: 'Sem permissões' }, 403)
  }

  await prisma.ordemItem.delete({ where: { id } })
  return c.json({ ok: true })
})

File: api/app/routers/ordem_items.py
import io
from datetime import UTC, datetime
from typing import Annotated, Any

from fastapi import APIRouter, Body, Depends, File, HTTPException, Query, UploadFile
from openpyxl import load_workbook  # type: ignore[import-untyped]
from sqlalchemy import func, select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.ordem_categories import (
    ItemShape,
    is_ordem_section,
    validate_item_data,
)
from app.core.ordem_permissions import ProfileForAuth, can_manage_item, resolve_category
from app.core.ordem_resolver import build_display, resolve_refs
from app.core.siie_atividades_import import map_activity_row
from app.db import get_session
from app.deps import AdminUser, CurrentUser
from app.models import OrdemItem, Profile, Scout
from app.models.enums import OrdemSection
from app.schemas.ordem_item import OrdemItemOut

router = APIRouter(prefix="/ordem-items", tags=["ordem-items"])


def _parse_dt(value: Any) -> datetime | None:
    if not isinstance(value, str) or not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    return parsed.astimezone(UTC).replace(tzinfo=None) if parsed.tzinfo else parsed


async def _load_auth(session: AsyncSession, user_id: str) -> ProfileForAuth | None:
    row = (
        await session.execute(
            select(Profile.role, Profile.roles, Profile.section).where(Profile.id == user_id)
        )
    ).first()
    if row is None:
        return None
    return ProfileForAuth(
        role=str(row.role), roles=list(row.roles), section=str(row.section) if row.section else None
    )


async def _validate_refs(
    session: AsyncSession, shape: ItemShape, value: dict[str, Any], section: str | None
) -> str | None:
    scout_ids: list[str] = []
    profile_ids: list[str] = []
    if shape == "MEMBER_REF" and isinstance(value.get("scoutId"), str):
        scout_ids.append(value["scoutId"])
    if shape == "NOITES_REF" and isinstance(value.get("scoutIds"), list):
        scout_ids.extend(i for i in value["scoutIds"] if isinstance(i, str))
    if shape == "PROFILE_REF" and isinstance(value.get("profileId"), str):
        profile_ids.append(value["profileId"])
    if shape == "SCOUT_OR_PROFILE_REF" and isinstance(value.get("refId"), str):
        if value.get("kind") == "scout":
            scout_ids.append(value["refId"])
        if value.get("kind") == "profile":
            profile_ids.append(value["refId"])

    if scout_ids:
        found = (
            await session.execute(
                select(Scout.id, Scout.section).where(Scout.id.in_(scout_ids))
            )
        ).all()
        if len(found) != len(scout_ids):
            return "Membro não encontrado"
        if section and any(s.section != section for s in found):
            return "O membro selecionado pertence a outra secção"
    if profile_ids:
        count = await session.scalar(
            select(func.count()).select_from(Profile).where(Profile.id.in_(profile_ids))
        )
        if count != len(profile_ids):
            return "Dirigente não encontrado"
    return None


def _annotated(item: OrdemItem, refs: Any) -> OrdemItemOut:
    out = OrdemItemOut.model_validate(item)
    out.data = {**(item.data or {}), "_display": build_display(item.data, refs)}
    return out


@router.get("")
async def list_items(
    user: CurrentUser,
    session: Annotated[AsyncSession, Depends(get_session)],
    from_: Annotated[str | None, Query(alias="from")] = None,
    to: Annotated[str | None, Query()] = None,
    section: Annotated[str | None, Query()] = None,
    category: Annotated[str | None, Query()] = None,
    included: Annotated[str | None, Query()] = None,
) -> dict[str, list[OrdemItemOut]]:
    conditions = []
    if from_:
        conditions.append(OrdemItem.date >= _require_dt(from_))
    if to:
        conditions.append(OrdemItem.date <= _require_dt(to))
    if section and is_ordem_section(section):
        conditions.append(OrdemItem.section == OrdemSection(section))
    if category:
        conditions.append(OrdemItem.category == category)
    if included == "true":
        conditions.append(OrdemItem.included_in_os_id.is_not(None))
    if included == "false":
        conditions.append(OrdemItem.included_in_os_id.is_(None))

    items = (
        await session.scalars(
            select(OrdemItem)
            .where(*conditions)
            .options(selectinload(OrdemItem.created_by))
            .order_by(OrdemItem.date.desc(), OrdemItem.created_at.desc())
        )
    ).all()
    refs = await resolve_refs(session, items)
    return {"items": [_annotated(item, refs) for item in items]}


def _require_dt(value: str) -> datetime:
    parsed = _parse_dt(value)
    if parsed is None:
        raise HTTPException(status_code=400, detail="Data inválida")
    return parsed


@router.post("")
async def create_item(
    user: CurrentUser,
    session: Annotated[AsyncSession, Depends(get_session)],
    body: Annotated[dict[str, Any], Body()],
) -> dict[str, OrdemItemOut]:
    raw_category = body.get("category")
    category = resolve_category(raw_category) if isinstance(raw_category, str) else None
    if category is None:
        raise HTTPException(status_code=400, detail="Categoria inválida")

    raw_section = body.get("section")
    if raw_section is None:
        section: str | None = None
    elif is_ordem_section(raw_section):
        section = raw_section
    else:
        raise HTTPException(status_code=400, detail="Secção inválida")
    if category.scope == "SECTION" and not section:
        raise HTTPException(status_code=400, detail="Secção é obrigatória")
    if category.scope == "GROUP" and section:
        raise HTTPException(status_code=400, detail="Categoria de grupo não aceita secção")

    date = _parse_dt(body.get("date"))
    if date is None:
        raise HTTPException(status_code=400, detail="Data inválida")

    result = validate_item_data(category.shape, body.get("data"))
    if not result.ok or result.value is None:
        raise HTTPException(status_code=400, detail=result.error or "Dados inválidos")

    profile = await _load_auth(session, user.id)
    if profile is None:
        raise HTTPException(status_code=404, detail="Perfil não encontrado")
    if not can_manage_item(profile, category, section):
        raise HTTPException(status_code=403, detail="Sem permissões para esta categoria/secção")

    ref_error = await _validate_refs(session, category.shape, result.value, section)
    if ref_error:
        raise HTTPException(status_code=400, detail=ref_error)

    item = OrdemItem(
        category=category.key,
        section=OrdemSection(section) if section else None,
        date=date,
        data=result.value,
        created_by_id=user.id,
    )
    session.add(item)
    await session.commit()
    created = await session.scalar(
        select(OrdemItem).where(OrdemItem.id == item.id).options(selectinload(OrdemItem.created_by))
    )
    assert created is not None
    return {"item": OrdemItemOut.model_validate(created)}


@router.post("/import-activities")
async def import_activities(
    admin: AdminUser,
    session: Annotated[AsyncSession, Depends(get_session)],
    file: Annotated[UploadFile | None, File()] = None,
) -> dict[str, Any]:
    if file is None:
        raise HTTPException(status_code=400, detail="Ficheiro não fornecido")
    try:
        content = await file.read()
        workbook = load_workbook(io.BytesIO(content), data_only=True, read_only=True)
        if not workbook.worksheets:
            raise HTTPException(status_code=400, detail="Folha vazia")
        sheet = workbook.worksheets[0]
        it = sheet.iter_rows(values_only=True)
        header = list(next(it, []))
        records = [
            dict(zip(header, row, strict=False)) for row in it if any(v is not None for v in row)
        ]
    except HTTPException:
        raise
    except Exception as err:  # noqa: BLE001
        raise HTTPException(
            status_code=400, detail=f"Não foi possível ler o ficheiro: {err}"
        ) from err

    summary: dict[str, Any] = {
        "total": len(records),
        "created": 0,
        "updated": 0,
        "skipped": 0,
        "errors": [],
    }

    valid = []
    for i, record in enumerate(records):
        mapped = map_activity_row(record, i + 2)
        if not mapped.ok:
            summary["errors"].append(
                {"row": mapped.row, "descricao": mapped.descricao, "error": mapped.error}
            )
        else:
            valid.append(mapped)

    if not valid:
        return {"summary": summary}

    ext_ids = [m.payload.external_id for m in valid if m.payload]
    existing_rows = (
        await session.execute(
            select(OrdemItem.external_id, OrdemItem.included_in_os_id).where(
                OrdemItem.external_id.in_(ext_ids)
            )
        )
    ).all()
    included_map = {r.external_id: r.included_in_os_id for r in existing_rows}

    for mapped in valid:
        payload = mapped.payload
        assert payload is not None
        was_existing = payload.external_id in included_map
        if was_existing and included_map[payload.external_id]:
            summary["skipped"] += 1
            continue
        data = {"nome": payload.nome, "datas": payload.datas, "local": payload.local}
        try:
            if was_existing:
                item = await session.scalar(
                    select(OrdemItem).where(OrdemItem.external_id == payload.external_id)
                )
                if item is not None:
                    item.category = "ATIVIDADE"
                    item.section = OrdemSection(payload.section) if payload.section else None
                    item.date = payload.date
                    item.data = data
                summary["updated"] += 1
            else:
                session.add(
                    OrdemItem(
                        external_id=payload.external_id,
                        category="ATIVIDADE",
                        section=OrdemSection(payload.section) if payload.section else None,
                        date=payload.date,
                        data=data,
                        created_by_id=admin.id,
                    )
                )
                summary["created"] += 1
            await session.commit()
        except IntegrityError:
            await session.rollback()
            if was_existing:
                summary["updated"] -= 1
            else:
                summary["created"] -= 1
            summary["errors"].append(
                {"row": mapped.row, "descricao": payload.nome, "error": "Conflito de unicidade"}
            )

    return {"summary": summary}


@router.patch("/{item_id}")
async def update_item(
    item_id: str,
    user: CurrentUser,
    session: Annotated[AsyncSession, Depends(get_session)],
    body: Annotated[dict[str, Any], Body()],
) -> dict[str, OrdemItemOut]:
    existing = await session.scalar(select(OrdemItem).where(OrdemItem.id == item_id))
    if existing is None:
        raise HTTPException(status_code=404, detail="Item não encontrado")
    if existing.included_in_os_id:
        raise HTTPException(status_code=409, detail="Item já incluído numa Ordem de Serviço")

    category = resolve_category(existing.category)
    if category is None:
        raise HTTPException(status_code=500, detail="Categoria inválida")

    profile = await _load_auth(session, user.id)
    if profile is None:
        raise HTTPException(status_code=404, detail="Perfil não encontrado")
    if existing.created_by_id != user.id and profile.role != "ADMIN":
        raise HTTPException(status_code=403, detail="Sem permissões")

    if "date" in body:
        date = _parse_dt(body["date"])
        if date is None:
            raise HTTPException(status_code=400, detail="Data inválida")
        existing.date = date
    if "data" in body:
        result = validate_item_data(category.shape, body["data"])
        if not result.ok or result.value is None:
            raise HTTPException(status_code=400, detail=result.error or "Dados inválidos")
        existing.data = result.value
    if "section" in body:
        if category.scope == "GROUP":
            raise HTTPException(status_code=400, detail="Categoria de grupo não aceita secção")
        if not is_ordem_section(body["section"]):
            raise HTTPException(status_code=400, detail="Secção inválida")
        if not can_manage_item(profile, category, body["section"]):
            raise HTTPException(status_code=403, detail="Sem permissões para essa secção")
        existing.section = OrdemSection(body["section"])

    await session.commit()
    updated = await session.scalar(
        select(OrdemItem)
        .where(OrdemItem.id == item_id)
        .options(selectinload(OrdemItem.created_by))
    )
    assert updated is not None
    return {"item": OrdemItemOut.model_validate(updated)}


@router.delete("/{item_id}")
async def delete_item(
    item_id: str, user: CurrentUser, session: Annotated[AsyncSession, Depends(get_session)]
) -> dict[str, bool]:
    existing = await session.scalar(select(OrdemItem).where(OrdemItem.id == item_id))
    if existing is None:
        raise HTTPException(status_code=404, detail="Item não encontrado")
    if existing.included_in_os_id:
        raise HTTPException(status_code=409, detail="Item já incluído numa Ordem de Serviço")
    profile = await _load_auth(session, user.id)
    if profile is None:
        raise HTTPException(status_code=404, detail="Perfil não encontrado")
    if existing.created_by_id != user.id and profile.role != "ADMIN":
        raise HTTPException(status_code=403, detail="Sem permissões")
    await session.delete(existing)
    await session.commit()
    return {"ok": True}
